### PR TITLE
Interface design of simple unified access to json, xml, yaml, ini, etc. 

### PR DESCRIPTION
**Pull request by lihuiba**
with examples and tests to show the usage

### CODE REVIEW

File: common/stream.h [hunk 1]
```diff
@@ -44,6 +45,19 @@ class IStream : public Object
...
+            void operator()(void* ptr) {
+                ::free(ptr);
+            }
+        };
+        std::unique_ptr<void, FreeDeleter> ptr;
+        ssize_t size;   // < 0 if error occured; |size| is always the # of bytes read
+    };
+
+    // read until EOF
+    ReadAll readall(size_t min_buf = 1024, size_t max_buf = 1024 * 1024 * 1024);
```
Comment (BigVan) [inline common/stream.h:59]: is the param 'min_buf' necessary?

File: common/simple_dom.h [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+// 1. simple & convenient usage;
+// 2. unified interface, common needs (may not fit for corner cases);
+// 3. efficient parsing (reading), as well as compiling;
+
+using str = estring_view;
+
+class RootNode;
+struct NodeWrapper;
+
+// the interface for internal implementations
```
Comment (liulanzheng) [inline common/simple_dom.h:49]: Can it be concealed for internal use and not disclosed externally?

File: common/simple_dom.h [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    bool valid() const { return _node; }
+    operator bool() const { return _node; }
+
+    size_t num_children() const {
+        return _node ? _node->num_children() : 0;
+    }
+    NodeWrapper get(size_t i) const {
+        IF_RET({_node->get(i)});
+    }
+    NodeWrapper get(str key) const {
```
Comment (liulanzheng) [inline common/simple_dom.h:152]: What is the result if the get operation fails?

File: common/simple_dom.h [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+inline NodeWrapper Node::wrapper() { return {this}; }
+
+// 1. text is moved to the simple_dom object, which frees it when destruct.
+// 2. the content of text may be modified in-place to un-escape strings.
+// 3. returning a pointer (of Node) is more efficient than an object (of Document),
+//    even if they are equivalent in binary form.
+Node* parse(char* text, size_t size, int flags);
+
+inline Node* parse(IStream::ReadAll&& buf, int flags) {
```
Comment (liulanzheng) [inline common/simple_dom.h:209]: Isn't the Node meant for internal use? How come it's being used by users here?

File: common/simple_dom.h [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+public:
+    void add_ref() { ++_refcnt; }
+    void del_ref()  { if (--_refcnt == 0) delete this; }
+};
+
+#define IF_RET(e) if (_node) return e; \
+                        else return {};
+
+// the interface for users
+struct NodeWrapper {
```
Comment (liulanzheng) [inline common/simple_dom.h:97]: For user use, it's better to be named as "Node". The original "Node" can be renamed to NodeBase or something else, and not be exposed externally.

Comment (lihuiba) [reply common/simple_dom.h:97]: OK

File: common/simple_dom.h (hunk 3 above)
Comment (lihuiba) [reply common/simple_dom.h:152]: an NodeWrapper with nullptr pointer to node implementation

File: common/simple_dom.h (hunk 4 above)
Comment (lihuiba) [reply common/simple_dom.h:209]: parse(...)->wrap()
```
// 3. returning a pointer (of Node) is more efficient than an object (of Document),
//    even if they are equivalent in binary form.
```

File: common/stream.h (hunk 1 above)
Comment (lihuiba) [reply common/stream.h:59]: in case that you want it start from, say, 1MB, to avoid some unnecessary realloc()

File: common/simple_dom.h (hunk 4 above)
Comment (liulanzheng) [reply common/simple_dom.h]: For users, it's a direct use after parsing, and calling wrap is a burden.

File: common/simple_dom.h (hunk 3 above)
Comment (liulanzheng) [reply common/simple_dom.h]: This requires a clear explanation and instructions for use.

File: common/simple_dom.h (hunk 4 above)
Comment (lihuiba) [reply common/simple_dom.h]: Not a big problem

Comment (liulanzheng) [reply common/simple_dom.h]: It a problem. The returned NodeImpl* is not the type used by the user and users must first determine if the returned NodeImpl* is nullptr before calling wrap. it's best not to expose these to users.

File: common/simple_dom.h (hunk 2 above)
Comment (lihuiba) [reply common/simple_dom.h]: No，because it's also the binary interface.

File: common/simple_dom.h (hunk 4 above)
Comment (lihuiba) [reply common/simple_dom.h]: OK

**Pull request merged**